packages/stack/deployment/test/deployment.spec.js

```diff
@@ -0,0 +1,186 @@
+      plugins._plugin.listeners[name].forEach(fn => {
+        fn(options, cb);
+      });
+    }
+  })
+};
+
+const fakeEmbark = {
+  events,
+  logger: {
+    info: () => {}
+  },
+  config: {
+    blockchainConfig: {}
+  },
+  assert: {
+    hasCommandHandler: (cmd) => {
+      sinon.assert.calledWith(fakeEmbark.events.setCommandHandler, cmd);
+    },
+    hasRequested: (cmd) => {
+      sinon.assert.calledWith(fakeEmbark.events.request, cmd);
+    },
+    hasRequestedWith: (cmd, ...args) => {
+      sinon.assert.calledWith(fakeEmbark.events.request, cmd, ...args);
+    },
+  },
+  teardown: () => {
+    fakeEmbark.events.listeners = {};
+    plugins._plugin.listeners = {};
+  }
+};
+
+describe('stack/deployment', () => {
+
+  let deployment;
+  let deployedContracts = [];
+
+  let beforeAllAction, beforeDeployAction, shouldDeployAction, deployedAction, afterAllAction;
+  let deployFn;
+  let doneCb;
+
+  beforeEach(() => {
+    deployment = new Deployment(fakeEmbark, { plugins });
+
+    beforeAllAction = sinon.spy((params, cb) => { cb(null, params); });
+    beforeDeployAction = sinon.spy((params, cb) => { cb(null, params); });
+    shouldDeployAction = sinon.spy((params, cb) => { cb(null, params); });
+    deployedAction = sinon.spy((params, cb) => { cb(null, params); });
+    afterAllAction = sinon.spy((params, cb) => { cb(null, params); });
+
+    deployFn = sinon.spy((contract, done) => {
+      deployedContracts.push(contract);
+      done(null, {}); // deployer needs to finish with a receipt object
+    });
+
+    doneCb = sinon.fake();
+  });
+
+  afterEach(() => {
+    deployedContracts = [];
+    fakeEmbark.teardown();
+    sinon.restore();
+  });
+
+  test('it should register deployFn and deploy contract by using it', () => {
+
+    let testContract = { className: 'TestContract', shouldDeploy: true };
+
+    plugins._plugin.registerActionForEvent('deployment:contract:beforeDeploy', beforeDeployAction);
+    plugins._plugin.registerActionForEvent('deployment:contract:shouldDeploy', shouldDeployAction);
+    plugins._plugin.registerActionForEvent('deployment:contract:deployed', deployedAction);
+
+    events.request('deployment:deployer:register', 'ethereum', deployFn);
+    events.request('deployment:contract:deploy', testContract, doneCb);
+
+    assert(beforeDeployAction.calledOnce)
+    assert(shouldDeployAction.calledOnce)
+    assert(deployFn.calledWith(testContract));
+    assert.equal(deployedContracts[0], testContract);
+    assert(doneCb.calledOnce);
+  });
+
+  test('it should deploy list of contracts', () => {
+
+    let testContracts = [
+      { className: 'Contract1', shouldDeploy: true },
+      { className: 'Contract2', shouldDeploy: true },
+      { className: 'Contract3', shouldDeploy: true }
+    ];
+
+    plugins._plugin.registerActionForEvent('deployment:deployContracts:beforeAll', beforeAllAction);
+    plugins._plugin.registerActionForEvent('deployment:contract:beforeDeploy', beforeDeployAction);
```

Comment:
This was interesting. When no `beforeDeployAction` is registered, the deployment process is stuck. All actions "registered for event" should be optional though.